### PR TITLE
feat: v2

### PR DESCRIPTION
**Pull request by bhumitattarde**

### CODE REVIEW

**Comment by rhnvrm** [comment]
@bhumitattarde, would it be possible to share a few steps to review this?

**Comment by bhumitattarde** [comment]
> @bhumitattarde, would it be possible to share a few steps to review this?

Skip the `Files Changed` tab and browse the branch directly. Focus on the `kitepp` directory. You've already looked at `responses/*` files so those can be skipped. Start with `kite.hpp`, it has interface of the `kite` (REST) class. The implementations can be found in the `kite/` directory. `ticker/ws.hpp` has interface for the websocket api and the implementation is in `internal.hpp`. After that, the only major files are `exceptions.hpp` and `utils.hpp`.

`tests` follows the similar structure.

**Comment by rhnvrm** [comment]
Thanks @bhumitattarde, gave a quick glance; will do a deeper dive over the weekend.

@ranjanrak @ajinasokan possible to review as well and also run the examples?

https://github.com/zerodha/cppkiteconnect/tree/v2

File: include/kitepp/kite/order.hpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        endpoint = "order.cancel";
+        fmtArgs = { variety, orderId };
+    };
+    return callApi<string, utils::json::JsonObject, true>(
+        endpoint, {}, fmtArgs, [](utils::json::JsonObject& data) {
+            return utils::json::get<string>(data, "order_id");
+        });
+};
+
+inline string kite::exitOrder(
```
Comment (ajinasokan) [inline include/kitepp/kite/order.hpp:96]: Is this for backward compatibility? If yes, since this is a major version change you could remove this alias function. And `exiting order` might confuse with `exiting position` as well, which is essentially placing a counter order.

File: include/kitepp/responses/market.hpp [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        open = getDouble(val[OPEN_IDX]);
+        high = getDouble(val[HIGH_IDX]);
+        low = getDouble(val[LOW_IDX]);
+        close = getDouble(val[CLOSE_IDX]);
+        volume = val[VOLUME_IDX].GetInt();
+        if (val.Size() > OI_IDX) { OI = val[OI_IDX].GetInt(); };
+    };
+
+    int volume = -1;
+    int OI = -1;
```
Comment (ajinasokan) [inline include/kitepp/responses/market.hpp:214]: Correct me if I'm wrong, `int` is 32 bit by default, irrespective of arch, right? Then these types should be changed to 64 bit. Better to change all instances of Volume and OI fields. This recently caused overflow issues in .Net lib. Ref: https://github.com/zerodha/dotnetkiteconnect/issues/34

File: include/kitepp/responses/mf.hpp [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    };
+
+    double minimumPurchaseAmount = -1;
+    double purchaseAmountMultiplier = -1;
+    double minimumAdditionalPurchaseAmount = -1;
+    double minimumRedemptionQuantity = -1;
+    double redemptionQuantityMultiplier = -1;
+    double lastPrice = -1;
+    bool purchaseAllowed = false;
+    bool redemtpionAllowed = false;
```
Comment (ajinasokan) [inline include/kitepp/responses/mf.hpp:276]: Typo: redemtpion -> redemption

File: include/kitepp/utils.hpp [hunk 4]
```diff
@@ -2,48 +2,600 @@
...
+    std::size_t end = 0;
+    while ((end = text.find(seperator, start)) != std::string::npos) {
+        tokens.emplace_back(text.substr(start, end - start));
         start = end + 1;
     };
 
     tokens.emplace_back(text.substr(start));
-
     return tokens;
 };
```
Comment (ajinasokan) [inline include/kitepp/utils.hpp:599]: Parsing CSV fields manually like this is bound to fail at some point, especially in financial data. Someday a company with a comma in the name like `Quora, Inc.` might show up in the instrument file. :)

I would suggest to use a standards compliant CSV parsing lib to parse fields.

File: include/kitepp/kite/internal.hpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+#include "../kite.hpp"
+#include "../utils.hpp"
+
+namespace kiteconnect {
+
+inline string kite::getAuth() const { return authorization; }
+
+inline string kite::encodeSymbolsList(const std::vector<string>& symbols) {
+    string str;
+    for (const auto& symbol : symbols) { str.append(FMT("i={0}&", symbol)); };
```
Comment (ajinasokan) [inline include/kitepp/kite/internal.hpp:40]: This `symbol` has to be URL encoded first. Otherwise it will fail for weird symbols like `M&M`.

File: examples/example2.cpp [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+
+        //! [initializing kite]
+        // kc::kite Kite(std::getenv("KITE_API_KEY"));
+        //! [initializing kite]
+
+        //! [obtaining login url]
+        // std::cout << "login URL: " << Kite.loginURL() << '\n';
+        // std::cout << "login with this URL and obtain the request token\n";
+        //! [obtaining login url]
+
```
Comment (ajinasokan) [inline examples/example2.cpp:49]: Are all these meant to be commented out? Is this for the docs?

**Comment by sukesh-ak** [comment]
Can you remove specific version dependency on `fmt` library. 
You are still on `fmt version 7.1.3` and current version `9.x` which has support for chrono and several other important options.



https://github.com/zerodha/cppkiteconnect/blob/ab0a5c96e8b4f72c59ffcc26aa530465f95e7738/include/kitepp/utils.hpp#L483

Here is the enum to string convertor for error messages (TODO item) which is throwing compile errors with latest fmt version
```c++

template <>
struct fmt::formatter<Error> : formatter<string_view>
{
  // parse is inherited from formatter<string_view>.
  template <typename FormatContext>
  auto format(Error e, FormatContext &ctx) const
  {
    string_view errstr = "unknown";
    switch (e)
    {
    case Error::Success: errstr = "Success"; break;
    case Error::Unknown: errstr = "Unknown"; break;
    case Error::Connection: errstr = "Connection"; break;
    case Error::BindIPAddress: errstr = "BindIPAddress"; break;
    case Error::Read: errstr = "Read"; break;
    case Error::Write: errstr = "Write"; break;
    case Error::ExceedRedirectCount: errstr = "ExceedRedirectCount"; break;
    case Error::Canceled: errstr = "Canceled"; break;
    case Error::SSLConnection: errstr = "SSLConnection"; break;
    case Error::SSLLoadingCerts: errstr = "SSLLoadingCerts"; break;
    case Error::SSLServerVerification: errstr = "SSLServerVerification"; break;
    case Error::UnsupportedMultipartBoundaryChars: errstr = "UnsupportedMultipartBoundaryChars"; break;
    case Error::Compression: errstr = "Compression"; break;
    }
    return formatter<string_view>::format(errstr, ctx);
  }
};
```

There are few bugs to be addressed related to fmt with latest version but nothing serious which blocks the upgrade to 9.x.

File: examples/example2.cpp (hunk 6 above)
Comment (bhumitattarde) [reply examples/example2.cpp:49]: Yes

File: include/kitepp/kite/internal.hpp (hunk 5 above)
Comment (bhumitattarde) [reply include/kitepp/kite/internal.hpp]: check now?

Comment (ajinasokan) [reply include/kitepp/kite/internal.hpp]: Looks alright

File: include/kitepp/utils.hpp (hunk 4 above)
Comment (bhumitattarde) [reply include/kitepp/utils.hpp]: Makes sense. I have fixed this.

File: include/kitepp/responses/market.hpp (hunk 2 above)
Comment (bhumitattarde) [reply include/kitepp/responses/market.hpp:215]: Will do.

Might be a good idea to mention precise types in the API docs too. Right now, [quotes section has volume type as just `int`](https://kite.trade/docs/connect/v3/market-quotes/#response-attributes). This led me to believe that volume is `int32` because `instrument_token` does have precise type mentioned (`uint32`).

**Review by ajinasokan — COMMENTED** [review]
LGTM

cc @rhnvrm @knadh

**Comment by bhumitattarde** [comment]
@rhnvrm @ajinasokan I don't have any more changes make. Can we merge this if there's nothing to add?

File: .git-blame-ignore-revs [hunk 7]
```diff
@@ -0,0 +1,2 @@
+# change column limit to 80
+39cd0478af477b63ecc925fa8af8b0c82e4566e6
```
Comment (rhnvrm) [inline .git-blame-ignore-revs:2]: Is this file needed? What does this do?

**Review by rhnvrm — APPROVED** [review]

File: .git-blame-ignore-revs (hunk 7 above)
Comment (bhumitattarde) [reply .git-blame-ignore-revs:2]: git blame ignores changes made by commits mentioned in this file. Mainly used to prevent git blame from getting trashed during mass-formatting commits.
